lib/propolis/src/accessors.rs

```diff
@@ -12,54 +12,84 @@
 //! contained emulation should fail any DMA accesses.
 
 use std::collections::{BTreeMap, BTreeSet, VecDeque};
+use std::ffi::c_void;
 use std::marker::PhantomData;
 use std::sync::{Arc, Mutex, MutexGuard, Weak};
 
 use crate::vmm::{MemCtx, VmmHdl};
 
-type AccId = usize;
-
-const ID_NULL: AccId = 0;
-const ID_ROOT: AccId = 1;
+/// Key type for identifying nodes referenced by `Tree`.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Copy, Clone)]
+#[repr(transparent)]
+pub struct NodeKey(*const Node<c_void>);
+impl NodeKey {
+    const NULL: Self = Self(std::ptr::null());
+}
+impl<T> From<&Arc<Node<T>>> for NodeKey {
+    fn from(value: &Arc<Node<T>>) -> Self {
+        NodeKey(Arc::as_ptr(value) as *const Node<c_void>)
+    }
+}
+impl std::fmt::Display for NodeKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:p}", self.0)
+    }
+}
+// Safety: While the key uses a pointer (!Send) type internally, it is for
+// unique identification purposes only, and is never meant to be dereferenced,
+// copied from, or transformed into a reference of any kind.
+unsafe impl Send for NodeKey {}
 
 struct TreeNode<T> {
-    parent_id: AccId,
+    /// [NodeKey] of the parent to this node
+    ///
+    /// Holds [NodeKey::NULL] if the node is the root of the [Tree]
+    parent_key: NodeKey,
+    /// [Weak] reference back to the node.  This access is needed if the node
+    /// undergoes adoption (being moved to a different [Tree])
     node_ref: Weak<Node<T>>,
-    children: BTreeSet<AccId>,
+    /// List of keys to child nodes (if any)
+    children: BTreeSet<NodeKey>,
```

Comment:
nit, not important: is the ordering provided by the `BTreeSet` load-bearing here? it might be nice if the comment also explained why.

Reply:
The ordering is not load-bearing at all.  I picked the btree version over hash due to this comment in the latter:

> In the current implementation, iterating over set takes O(capacity) time instead of O(len) because it internally visits empty buckets too.

I don't plan to comment on the choice, since it's not an important detail of the implementation. (If that seems alright)

Reply:
fine with me!